File: BddPipe/BddPipe/Runner.cs
using System;
using System.Runtime.ExceptionServices;
using System.Threading.Tasks;
using BddPipe.Model;

namespace BddPipe
{
    /// <summary>
    /// BddPipe scenario runner
    /// </summary>
    public static partial class Runner
    {
        private static Pipe<R> RunStepCommon<T, R>(this in Pipe<T> pipe,
                                                        Some<Title> title,
                                                        in Either<Func<T, R>, Func<T, Task<R>>> stepFunc) =>
            Execute(
                pipe,
                stepFunc,
                (source, fn) => ProcessStep(title, source, fn),
                (source, fn) => ProcessStep(title, source, fn),
                (source, fn) => ProcessStep(title, source, fn),
                (source, fn) => ProcessStep(title, source, fn)
            );

        private static Either<Ctn<ExceptionDispatchInfo>, Ctn<R>> ToStepResult<T, R>(this Result<R> result, Ctn<T> tValue, Some<Title> title) =>
            result.Match<Either<Ctn<ExceptionDispatchInfo>, Ctn<R>>>(
                r => tValue.ToCtn(r, title.ToStepOutcome(Outcome.Pass)),
                ex => tValue.ToCtn(ex, title.ToStepOutcome(new Some<Exception>(ex.SourceException).ToOutcome()))
            );

        private static Either<Ctn<ExceptionDispatchInfo>, Ctn<R>> ToStepErrorState<R>(this Ctn<ExceptionDispatchInfo> err, in Some<Title> title) =>
            err.ToCtn(err.Content, title.ToStepOutcome(Outcome.NotRun));

        private static Either<Ctn<ExceptionDispatchInfo>, Ctn<R>> ProcessStep<T, R>(
            Some<Title> title,
            in Either<Ctn<ExceptionDispatchInfo>, Ctn<T>> source,
            Func<T, R> step)
        {
            return source.Match(
                tValue => step
                    .Apply(tValue.Content)
                    .TryRun()
                    .ToStepResult(tValue, title),
                err => err.ToStepErrorState<R>(title)
            );
        }

        private static Task<Either<Ctn<ExceptionDispatchInfo>, Ctn<R>>> ProcessStep<T, R>(
            Some<Title> title,
            in Either<Ctn<ExceptionDispatchInfo>, Ctn<T>> source,
            Func<T, Task<R>> step)
        {
            return source.Match(
                async tValue => (await step
                        .Apply(tValue.Content)
                        .TryRunAsync().ConfigureAwait(false))
                        .ToStepResult(tValue, title),
                err => Task.FromResult(err.ToStepErrorState<R>(title)));
        }

        private static async Task<Either<Ctn<ExceptionDispatchInfo>, Ctn<R>>> ProcessStep<T, R>(
            Some<Title> title,
            Task<Either<Ctn<ExceptionDispatchInfo>, Ctn<T>>> source,
            Func<T, Task<R>> step)
        {
            var sourceInstance = await source.ConfigureAwait(false);
            return await ProcessStep(title, sourceInstance, step).ConfigureAwait(false);
        }

        private static async Task<Either<Ctn<ExceptionDispatchInfo>, Ctn<R>>> ProcessStep<T, R>(
            Some<Title> title,
            Task<Either<Ctn<ExceptionDispatchInfo>, Ctn<T>>> source,
            Func<T, R> step)
        {
            var sourceInstance = await source.ConfigureAwait(false);
            return ProcessStep(title, sourceInstance, step);
        }

        private static Pipe<R> RunStep<T, R>(this Pipe<T> pipe, Some<Title> title, Func<T, Task<R>> step) =>
            pipe.RunStepCommon<T, R>(title, step);

        private static Pipe<R> RunStep<T, R>(this Pipe<T> pipe, Some<Title> title, Func<T, R> step) =>
            pipe.RunStepCommon<T, R>(title, step);
    }
}

File: BddPipe/BddPipe.FSharp/Runner.fs
namespace BddPipe.FSharp

open BddPipe.FSharp.Model
open BddPipe.FSharp.StepResultExtensions
open BddPipe.FSharp.CtnExtensions
open System.Runtime.ExceptionServices

module Runner =

    let private toStepResult<'t, 'r> (tValue: Ctn<'t>, title: Title) (result: Result<'r, ExceptionDispatchInfo>) =
        match result with
            | Ok r ->
                CtnOk (tValue |> toCtn (r, title |> toStepOutcome Pass))
            | Error ex ->
                CtnError (tValue |> toCtn (ex, title |> toStepOutcome (toOutcome ex.SourceException)))

    let private toStepErrorState<'t>(title: Title) (ctnError: Ctn<ExceptionDispatchInfo>) : EitherCtn<'t> =
        CtnError (ctnError |> toCtn(ctnError.Content, title |> toStepOutcome NotRun))

    let private tryRun<'t> (fn: unit -> 't) =
        try
            let result = fn()
            Ok result
        with | ex ->
            Error (ExceptionDispatchInfo.Capture(ex))

    let private tryRunAsync<'t> (fn: unit -> Async<'t>) =
        async {
            try
                let! result = fn()
                return Ok result
            with | ex ->
                return Error (ExceptionDispatchInfo.Capture(ex))
        }

    let private processStep<'t,'r> (title: Title, source: EitherCtn<'t>, stepFn: 't -> 'r) : EitherCtn<'r> =
        match source with
            | CtnOk tValue ->
                tryRun (fun () -> stepFn tValue.Content)
                |> toStepResult (tValue, title)
            | CtnError err -> err |> toStepErrorState<'r> title

    let private processStepAsync<'t, 'r> (title: Title, source: EitherCtn<'t>, stepFn: 't -> Async<'r>) : Async<EitherCtn<'r>> =
        async {
            match source with
                | CtnOk tValue ->
                    let! result = tryRunAsync (fun () -> stepFn tValue.Content)
                    return result |> toStepResult (tValue, title)
                | CtnError err -> return err |> toStepErrorState<'r> title
        }
